test/icinga-checkresult.cpp

```diff
@@ -809,4 +814,183 @@ BOOST_AUTO_TEST_CASE(service_flapping_ok_over_bad_into_ok)
 
 #endif /* I2_DEBUG */
 }
+
+BOOST_AUTO_TEST_CASE(suppressed_notification)
+{
+	/* Tests that suppressed notifications on a Checkable are sent after the suppression ends if and only if the first
+	 * hard state after the suppression is different from the last hard state before the suppression. The test works
+	 * by bringing a service in a defined hard state, creating a downtime, performing some state changes, removing the
+	 * downtime, bringing the service into another defined hard state (if not already) and checking the requested
+	 * notifications.
+	 */
+
+	const std::vector<ServiceState> states {ServiceOK, ServiceWarning, ServiceCritical, ServiceUnknown};
+
+	for (bool isVolatile : {false, true}) {
+		for (int checkAttempts : {1, 2}) {
+			for (ServiceState initialState: states) {
+				for (auto s1 : states) for (auto s2 : states) for (auto s3 : states) for (auto s4 : states) {
```

Comment:
Holy mother of room and time, you've just beat me (me!) in 🍝 code!

Feel free to ignore { } and indentation, but put them one per line.

Reply:
Well, temporarily I had code in there to dynamically generate the sequence. But I removed it as I thought it was overkill if you can write this as a few nested loops. I wish STL had something like `std::next_permutation()` for Cartesian products...

Reply:
I consider

```
for (auto s1 : states)
for (auto s2 : states)
for (auto s3 : states)
for (auto s4 : states) {
```

fine.